### DISCUSSION TITLE
Configuring Runestone Server for local development

### DISCUSSION THREAD

**Discussion by cfcurtis** [MRU-COMP1701-dev/fopp-pretext#1] [General]
Here's what I've done so far (incomplete):

The goal is to first convert to PreTeXt, then modify for COMP 1701 at MRU.

Author guide: https://runestone.academy/ns/books/published/authorguide/developer.html?mode=browsing
- Somewhat out of date, new books should be developed as PreTeXt
- Conversion utility: https://github.com/bnmnetp/Runestone2PreTeXt

## Steps actually taken
- Forked https://github.com/RunestoneInteractive/fopp as https://github.com/mru-comp1701-dev/fopp-pretext
- Followed "Getting Started" directions on the [book README](https://github.com/RunestoneInteractive/fopp) to install runestone
	- Used Python 3.11.1 in a venv on Fedora 37
	- Local build resulted in the following: `RuntimeError: Cannot configure a Database URL!`
	- This can be entirely ignored and `runestone serve` will serve the book at http://localhost:8000/
	- This is a preview only and doesn't actually enable interactive components
- To try to get interactive components, I installed the Runestone server following the Docker instructions at https://runestoneserver.readthedocs.io/en/latest/docker/README.html

### Attempted rst to ptx conversion
- Cloned https://github.com/bnmnetp/Runestone2PreTeXt conversion tool
- Following README instructions:
    1. Converted rst to PreTexT, but first had to edit the [main runestone script](https://github.com/RunestoneInteractive/RunestoneComponents/blob/master/runestone/__main__.py) at line 343 to check for the DBURL key before deleting it. This successfully spat out a bunch of xml files, albeit with a bunch of warnings.
    2. Ran xml2ptx.py, which choked every time a `<` showed up in sample code. These files were skipped but others were successfully converted and placed in pretext/Chapters.
    3. Ran the various fix scripts, not sure what these did. Reformat had four failures:
    ```
    Failed to reformat pretext/AdvancedAccumulation/ChapterAssessment.ptx
    Failed to reformat pretext/FrontBackMatter/copyright.ptx
    Failed to reformat pretext/Inheritance/inheritVarsAndMethods.ptx
    Failed to reformat pretext/SimplePythonData/Input.ptx
     ```
     4. TBD: creating PreTeXt project file, updating TOC, building book, etc. Step 2 probably needs work first though.

**Comment by matthew-hre** [comment]
Yep, ran into the same `KeyError: 'DBURL'` error, commented out line 343 in `runestone/__main__.py` to resolve.

### Some notes from my explorations so far:

- Instead of using `xml2ptx.py`, went with the `xsltproc /docutils2ptx.xsl /build/xml/{page}.xml > {page}.ptx` route, and encountered no errors (although I only tried a few pages).
- Had to migrate the toctree directives over to `main.ptx`, for future reference, ended up looking like this:
  ```xml
  <toc maxdepth="2">
    <caption>Contents:</caption>
  </toc>
    ```
- Attempting to compile the web build crashed with a wall of the following: `PTX:BUG:   the $chunk-level variable is empty, so must have been set improperly, perhaps when overridden in a conversion.  The "containing-filename" template is likely to act incorrectly.  Please report me`. Will I report that? Not yet.

Will attempt this process again from a fresh clone, as to not get distracted while having to pip-install every thirty seconds.

**UPDATE**: Hit the same error wall from a fresh clone as well. Managed to get xml2ptx working by modifying the `xsl_filename` variable.
